### ISSUE TITLE
feature request: allow outcomes outside the range [0-1] in reg:logistic

### ISSUE THREAD

**Issue by osofr** [dmlc/xgboost#2118]
There are some situations when the outcome might fall outside the range of (0,1), yet we still would like to run the regression based on log-likelihood loss function for binary outcomes (i.e., the logistic regression). For example, this might be useful to ensure certain modeling constraints on our regression model, while avoiding arbitrary outcome rescaling decisions. While manually rescaling the outcome back to [0-1] might not be appropriate for these applications, the logistic regression loss function is still fully applicable for such regression problems. This should work just fine and give correct results, especially if the loss / gradient are implemented correctly (and given that the boundary cases such as 0*log(0) are taken care of). 

Is there any interest in enabling this feature in xgboost? I have removed the relevant error check https://github.com/dmlc/xgboost/blob/master/src/objective/regression_obj.cc#L34
and recompiled the package and everything appears to be working correctly. That is, the logistic regression runs fine for outcomes outside [0-1], while the modeling predictions all fall within the range of [0-1]. All of this is exactly what would be expected from correctly implemented logistic loss/gradient. If there is any interest in adding this, please let me know, I'd be happy to do a pull request. 

Best,
Oleg
